proxy/grpcproxy/cluster.go

```diff
@@ -15,38 +15,177 @@
+		watcher, err := cp.gr.Resolve(prefix)
+		if err != nil {
+			plog.Warningf("failed to resolve %q (%v)", prefix, err)
+			continue
+		}
+		cp.wa = watcher
+		break
+	}
+
+	// pass 'donec' to signal monitor routine return
+	go cp.monitor(donec)
 }
 
-func NewClusterProxy(c *clientv3.Client) pb.ClusterServer {
-	return &clusterProxy{
-		client: c,
+func (cp *clusterProxy) monitor(donec chan struct{}) {
+	// to signal monitor routine return
+	defer close(donec)
+
+	for cp.ctx.Err() == nil {
+		ups, err := cp.wa.Next()
+		if err != nil {
+			plog.Warningf("clusterProxy watcher error (%v)", err)
+			if grpc.ErrorDesc(err) == naming.ErrWatcherClosed.Error() {
+				return
+			}
+		}
+
+		cp.umu.Lock()
+		if len(cp.umap) == 0 {
+			cp.umap = make(map[string]struct{}, len(ups))
+		}
+		for i := range ups {
+			switch ups[i].Op {
+			case gnaming.Add:
+				cp.umap[ups[i].Addr] = struct{}{}
+			case gnaming.Delete:
+				delete(cp.umap, ups[i].Addr)
+			}
+		}
+		cp.umu.Unlock()
+	}
+}
+
+func (cp *clusterProxy) members() []*pb.Member {
+	cp.umu.RLock()
+	defer cp.umu.RUnlock()
+
+	// using its own listen client URLs
+	// or addresses haven't been populated from the watcher
+	if cp.self || len(cp.umap) == 0 {
+		return []*pb.Member{{ClientURLs: []string{cp.laddr}}}
+	}
+
+	mbs := make([]*pb.Member, 0, len(cp.umap))
+	for addr := range cp.umap {
+		mbs = append(mbs, &pb.Member{ClientURLs: []string{addr}})
 	}
+	return mbs
 }
 
 func (cp *clusterProxy) MemberAdd(ctx context.Context, r *pb.MemberAddRequest) (*pb.MemberAddResponse, error) {
-	conn := cp.client.ActiveConnection()
-	return pb.NewClusterClient(conn).MemberAdd(ctx, r)
+	var ps []string
+	if r != nil {
+		ps = r.PeerURLs
+	}
+	mresp, err := cp.clus.MemberAdd(ctx, ps)
+	if err != nil {
+		return nil, err
+	}
+	resp := (pb.MemberAddResponse)(*mresp)
+	return &resp, err
 }
 
 func (cp *clusterProxy) MemberRemove(ctx context.Context, r *pb.MemberRemoveRequest) (*pb.MemberRemoveResponse, error) {
-	conn := cp.client.ActiveConnection()
-	return pb.NewClusterClient(conn).MemberRemove(ctx, r)
+	var id uint64
+	if r != nil {
+		id = r.ID
+	}
+	mresp, err := cp.clus.MemberRemove(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	resp := (pb.MemberRemoveResponse)(*mresp)
+	return &resp, err
 }
 
 func (cp *clusterProxy) MemberUpdate(ctx context.Context, r *pb.MemberUpdateRequest) (*pb.MemberUpdateResponse, error) {
-	conn := cp.client.ActiveConnection()
-	return pb.NewClusterClient(conn).MemberUpdate(ctx, r)
+	var id uint64
+	var ps []string
+	if r != nil {
+		id = r.ID
+		ps = r.PeerURLs
+	}
+	mresp, err := cp.clus.MemberUpdate(ctx, id, ps)
+	if err != nil {
+		return nil, err
+	}
+	resp := (pb.MemberUpdateResponse)(*mresp)
+	return &resp, err
 }
 
 func (cp *clusterProxy) MemberList(ctx context.Context, r *pb.MemberListRequest) (*pb.MemberListResponse, error) {
-	conn := cp.client.ActiveConnection()
-	return pb.NewClusterClient(conn).MemberList(ctx, r)
+	if cp.self || cp.wa != nil {
```

Comment:
maybe `if cp.umap != nil`? if umap is never initialized then it's configured to passthrough